pkg/kubelet/kubeletconfig/checkpoint/download.go

```diff
@@ -60,47 +64,44 @@ type RemoteConfigSource interface {
 	object() interface{}
 }
 
-// NewRemoteConfigSource constructs a RemoteConfigSource from a v1/NodeConfigSource object, or returns
-// a sanitized failure reason and an error if the `source` is blatantly invalid.
+// NewRemoteConfigSource constructs a RemoteConfigSource from a v1/NodeConfigSource object
 // You should only call this with a non-nil config source.
+// Note that the API server validates Node.Spec.ConfigSource.
 func NewRemoteConfigSource(source *apiv1.NodeConfigSource) (RemoteConfigSource, string, error) {
-	// exactly one subfield of the config source must be non-nil, toady ConfigMapRef is the only reference
-	if source.ConfigMapRef == nil {
-		return nil, status.FailSyncReasonAllNilSubfields, fmt.Errorf("%s, NodeConfigSource was: %#v", status.FailSyncReasonAllNilSubfields, source)
-	}
-
-	// validate the NodeConfigSource:
-
-	// at this point we know we're using the ConfigMapRef subfield
-	ref := source.ConfigMapRef
+	// NOTE: Even though the API server validates the config, we check whether all *known* fields are
+	// nil here, so that if a new API server allows a new config source type, old clients can send
+	// an error message rather than crashing due to a nil pointer dereference.
 
-	// name, namespace, and UID must all be non-empty for ConfigMapRef
-	if ref.Name == "" || ref.Namespace == "" || string(ref.UID) == "" {
-		return nil, status.FailSyncReasonPartialObjectReference, fmt.Errorf("%s, ObjectReference was: %#v", status.FailSyncReasonPartialObjectReference, ref)
+	// exactly one reference subfield of the config source must be non-nil, today ConfigMap is the only reference subfield
+	if source.ConfigMap == nil {
+		return nil, status.FailSyncReasonAllNilSubfields, fmt.Errorf("%s, NodeConfigSource was: %#v", status.FailSyncReasonAllNilSubfields, source)
 	}
-
 	return &remoteConfigMap{source}, "", nil
 }
 
 // DecodeRemoteConfigSource is a helper for using the apimachinery to decode serialized RemoteConfigSources;
 // e.g. the metadata stored by checkpoint/store/fsstore.go
 func DecodeRemoteConfigSource(data []byte) (RemoteConfigSource, error) {
 	// decode the remote config source
-	obj, err := runtime.Decode(legacyscheme.Codecs.UniversalDecoder(), data)
+	_, codecs, err := scheme.NewSchemeAndCodecs()
+	if err != nil {
+		return nil, err
+	}
+
+	obj, err := runtime.Decode(codecs.UniversalDecoder(), data)
 	if err != nil {
 		return nil, fmt.Errorf("failed to decode, error: %v", err)
 	}
 
-	// for now we assume we are trying to load an apiv1.NodeConfigSource,
+	// for now we assume we are trying to load an kubeletconfigv1beta1.SerializedNodeConfigSource,
```

Comment:
should this be `kubeletconfiginternal.SerializedNodeConfigSource`?

Reply:
No. The external version is what we load; the decoder automatically converts it to the internal version as part of the decoding process.